src/sinks/prometheus/exporter.rs

```diff
@@ -302,175 +434,93 @@ impl PrometheusExporter {
+                let mut metrics = self.metrics.write().unwrap();
+
+                let metric_ref = MetricRef::from_metric(&normalized);
+                match metrics.get_mut(&metric_ref) {
+                    Some((data, metadata)) => {
+                        *data = normalized;
+                        metadata.last_updated = Instant::now();
+                    }
+                    None => {
+                        // We specifically clear sets when they expire rather than totally delete
+                        // them, and this flag is how we know to do that.
+                        let should_clear = matches!(normalized.value(), MetricValue::Set { .. });
+                        metrics.insert(metric_ref, (normalized, MetricMetadata::new(should_clear)));
+                    }
+                }
             }
 
-            match item.kind() {
-                MetricKind::Incremental => {
-                    let mut entry = MetricEntry(item.into_absolute());
-                    if let Some((MetricEntry(mut metric), _)) = metrics.map.remove_entry(&entry) {
-                        if metric.update(&entry) {
-                            entry = MetricEntry(metric);
-                        } else {
-                            warn!(message = "Metric changed type, dropping old value.", series = %entry.series());
+            // If we've exceed our flush interval, go through all of the metrics we're currently
+            // tracking and remove any which have exceeded the the flush interval in terms of not
+            // having been updated within that long of a time.
+            //
+            // TODO: Can we be smarter about this? As is, we might wait up to 2x the flush period to
+            // remove an expired metric depending on how things line up.  It'd be cool to _check_
+            // for expired metrics more often, but we also don't want to check _way_ too often, like
+            // every second, since then we're constantly iterating through every metric, etc etc.
+            if last_flush.elapsed() > self.config.flush_period_secs {
+                last_flush = Instant::now();
+
+                let mut metrics = self.metrics.write().unwrap();
+
+                let metrics_to_expire = metrics
+                    .iter()
+                    .filter(|(_, (_, metadata))| {
+                        !metadata.has_expired(self.config.flush_period_secs)
+                    })
+                    .map(|(metric_ref, (_, metadata))| (metric_ref.clone(), *metadata))
+                    .collect::<Vec<_>>();
+
+                for (metric_ref, metadata) in metrics_to_expire {
+                    if metadata.should_clear {
+                        if let Some((metric, _)) = metrics.get_mut(&metric_ref) {
+                            metric.data_mut().value_mut().zero();
                         }
+                    } else {
+                        metrics.remove(&metric_ref);
                     }
```

Comment:
Can this be rewritten using the `retain` interface to avoid creating the temporary `metrics_to_expire`?

Reply:
We'd still have to iterate over the unretained metrics to see if we needed to zero them out, so I think it's a wash either way.

What we really need is for `HashMap::drain_filter` to be stabilized so we can capture the expired metrics and then be able to use them to do the "should we clear this metric?" logic.